hsmd: Add hsmd_{new,ready}_channel

ksedgwic opened the pull request:
These messages are used by external signers which need to maintain channel state.

- [x] Consider `dualopend` support

m-schmoock committed:
pytest: fix test_pluginconnected_hook_chaining

The last line of the testcase was checking on the wrong node l3
instead of l1. l3 didn't had the plugins configured that would
produce the log entry we were looking for not to be present.

Changelog-None

jsarenik committed:
Set default port according to network

The idea is to have different default ports for different networks.

Current default port is `9735` for everything. Let's use it for
the mainnet and reuse the difference added to the default port
from `rpc_port` values in `bitcoin/chainstate.c`.

Testnet would be `19735` (adding rpc_port - 8332 = `10000`).

Signet would be `39735` (adding rpc_port - 8332 = `30000`).

Regtest would be `19846` (adding rpc_port - 8332 = `10111`).

With Vincenzo's kind pair-programming help over tmate.

Two other commits were squashed into this one so that bisecting
never ends up in half-baked state:

1. chainparams: Fix regtest default rpc_port

   bitcoind -help says this:

    -rpcport=<port>
         Listen for JSON-RPC connections on <port> (default: 8332, testnet:
         18332, signet: 38332, regtest: 18443)

2. test_gossip: Default port for regtest

   hex: 2607 is now .... (could be 4d86 but Elements uses another port)
   dec: 9735 is now any port (could be 19846 ^^ but now is for any port)

   The lines which were binding to default port were removed as the
   default port is different on each network.

NOTE: Remember not to modify gossip_store tests which loads everything raw
      including the checksums.

Changelog-Changed: If the port is unspecified, the default port is chosen according to used network similarly to Bitcoin Core.

ksedgwic committed:
hsmd: Add hsmd_new_channel

ksedgwic committed:
hsmd: Add hsmd_ready_channel

johnongit committed:
Upgrade bitcoin-cli to 22.0

andrewtoth committed:
refactor: move exclude parsing to json_tok

jsarenik committed:
doc: lightningd-config.5.md - default ports

Also nit: s/possible/possibly suggested by michaelfolkson.

devrandom commented on hsmd/hsmd_wire.csv:
```diff
@@ -23,6 +23,16 @@ msgdata,hsmd_init_reply,bip32,ext_key,
 
+# Declare a new channel.
+msgtype,hsmd_new_channel,30
+# Which identity to use for requests
```
I don't see fields commented in other messages in this file, so maybe remove these comments for style consistency

ksedgwic replied:
done

devrandom commented on hsmd/libhsmd.c:
```diff
@@ -276,6 +280,76 @@ static void get_channel_seed(const struct node_id *peer_id, u64 dbid,
+
+	/* Stub implementation */
+
+	/* Fail fast if any values are obviously uninitialized. */
```
maybe "uninitialized or otherwise obviously wrong"

ksedgwic replied:
done

devrandom reviewed (approved):
LGTM

andrewtoth committed:
Add route exclusion payment modifier and pay rpc exclude arg 

Changelog-Added: Add `exclude` option for `pay` command to manually exclude channels or nodes when finding a route.

andrewtoth committed:
tests: add pay test for exclude arg

andrewtoth committed:
doc: add exclude arg doc for pay rpc

ZmnSCPxj committed:
wallet/db.c, wallet/wallet.c: Add a partial index to speed up startup.

Closes: #4901

Tested by `EXPLAIN QUERY PLAN` on sqlite3; #4901 shows the result from
@whitslack doing a similar partial index on PostgreSQL on his ~1000 chan
node.

ChangeLog-Added: db: Speed up loading of pending HTLCs during startup by using a partial index.

NicolasDorier committed:
Fix dockerfile for arm32/64

rustyrussell committed:
pytest: fix flake in testing.

As noted in https://github.com/ElementsProject/lightning/pull/4897/commits/0a406230d051a83f9f029054efe2dd24e87887a8#diff-5871d4c569454db5e625383975462132da0bd03d32df145d8d72d8fafd86d952R3544-R3546

Turns out we sometimes hang up before l2 sees the previous tx revoked,
so we get a normal unilateral close, not a cheat.

Reported-by: Simon Vrouwe
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

devrandom commented on lightningd/dual_open_control.c:
```diff
@@ -645,13 +645,23 @@ openchannel2_hook_cb(struct openchannel2_payload *payload STEALS)
+	// Determine the wallet index for our upfront_shutdown_script, UINT32_MAX if not found.
+	u32 our_shutdown_script_wallet_index = UINT32_MAX;
+	bool is_p2sh;
+	wallet_can_spend(
```
here and elsewhere, shouldn't we do something special if `wallet_can_spend()` returns false?

devrandom replied:
Well, actually it's fine as it is, because it's not an error if it's not in the wallet - it may be whitelisted.

devrandom commented on lightningd/opening_common.c:
```diff
@@ -3,13 +3,16 @@
 #include <common/type_to_string.h>
 #include <common/wire_error.h>
 #include <connectd/connectd_wiregen.h>
+#include <errno.h>
```
how is errno.h used?

ksedgwic replied:
```
	/* Declare the new channel to the HSM. */
	new_channel_msg = towire_hsmd_new_channel(NULL, &uc->peer->id, uc->dbid);
	if (!wire_sync_write(ld->hsm_fd, take(new_channel_msg)))
		fatal("Could not write to HSM: %s", strerror(errno));
```

devrandom commented on openingd/dualopend.c:
```diff
@@ -2408,9 +2435,40 @@ static u8 *opener_commits(struct state *state,
 
+	if (!amount_sat_to_msat(&their_msats, tx_state->accepter_funding)) {
+		status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			      "Rounding error, can't convert accepter_funding %s"
```
it's an overflow error, not a rounding error

ksedgwic replied:
fixed

devrandom commented:
re-LGTM

rustyrussell committed:
pytest: test to show conflict between websocket and wildcard addresses.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
connectd: fix websocket binding when we're doing both IPv4 and IPv6 on same port.

We would fail connectd when listening on the IPv6 version failed; instead we should
allow that.

Changelog-Experimental: experimental-websocket-port fixed to work with default addresses.
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
options: fix handling of wildcard (allproto) address.

We treated ':' as an empty DNS name in EXPERIMENTAL, which is wrong.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

ZmnSCPxj committed:
doc/lightning-datastore.7.md: Fix typos.

Changelog-None

jsarenik committed:
.github: Fix FreeBSD workflow

The docs https://github.com/marketplace/actions/freebsd-vm
say that `runs-on:` must be `macos-10.15`.

m-schmoock committed:
status: do not log ping and pong packets

Debug logging of ping and pong packets turned out to be
too excessive since this is now done on all channels
and without config option.

Changelog-None

andrewtoth committed:
make: remove generated files when running clean

rustyrussell committed:
connectd: don't advertize websocket support if we have no other (non-Tor) addrs.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
check: make sure all files outside contrib/ include "config.h" first.

And turn "" includes into full-path (which makes it easier to put
config.h first, and finds some cases check-includes.sh missed
previously).

config.h sets _GNU_SOURCE which really needs to be done before any
'#includes': we mainly got away with it with glibc, but other platforms
like Alpine may have stricter requirements.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
common: remove unused io_lock.c

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
tools/check-includes.sh: make shellcheck happy.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
devtools: simple script to look for functions which don't seem to be used.

Gives a rough first-pass, at least.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
bitcoin: remove unused functions, or make static.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
common: remove unused functions or make static.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
daemons: remove unused functions or make static.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
gossipd: remove gossip_msg.[ch]

This was a remnant from when we used to get routing from gossipd.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
pytest: fix occasional flake in test_announce_address

We make sure the gossip msg was sent, but the other node might not
have digested it yet:

```
        # Check other node can parse these
>       addresses = l2.rpc.listnodes(l1.info['id'])['nodes'][0]['addresses']
E       KeyError: 'addresses'
```

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

whitslack committed:
common/utils: introduce tmpdir_mkstemp; use it

Various unit tests were creating temporary files unconditionally in /tmp
and were not cleaning up after themselves. Introduce a new variant of
mkstemp(3p) that respects the TMPDIR environment variable, and use it in
the offending unit tests. This allows each test run to use a dedicated
TMPDIR that can be cleaned up after the run.

Changelog-None

Signed-off-by: Matt Whitlock <c-lightning@mattwhitlock.name>

m-schmoock committed:
connectd: fix empty error message

1. Adds the missing DNS error massages so they can be handled by
   connect_control.
2. Prepends a 'All addresses failed' to code 401 message, so we
   always have at least some error message to the user.

Changelog-None

ksedgwic committed:
wallet: make wallet_can_spend non-static because needed

ksedgwic commented:
This PR was superseded by upstream submission:
https://github.com/ElementsProject/lightning/pull/4949
